main.rb

```diff
@@ -0,0 +1,186 @@
+# rubocop:disable Metrics/ModuleLength
+# rubocop:disable Metrics/CyclomaticComplexity
+# rubocop:disable Metrics/PerceivedComplexity
+
+module Enumerable
+  def my_each
+    return enum_for(:my_each) unless block_given?
+
+    new_self = self
+    i = 0
+    new_self = new_self.instance_of?(Array) || new_self.instance_of?(Hash) ? flatten : to_a
+    if instance_of?(Array) || instance_of?(Range)
+      new_self.length.times do
+        yield(new_self[i])
+        i += 1
+      end
+    elsif instance_of?(Hash)
+      new_self.length / 2.times do
+        yield(new_self[i], new_self[i + 1])
+        i += 2
+      end
+    end
+    new_self
+  end
+
+  def my_each_with_index
+    return enum_for(:my_each_with_index) unless block_given?
+
+    new_self = self
+    i = 0
+    new_self = new_self.instance_of?(Array) || new_self.instance_of?(Hash) ? flatten : to_a
+    new_self.length.times do
+      yield(new_self[i], i)
+      i += 1
+    end
+    new_self
+  end
+
+  def my_select
+    return enum_for(:my_select) unless block_given?
+
+    i = 0
+    arr = []
+    new_self = self
+    new_self = new_self.instance_of?(Array) || new_self.instance_of?(Hash) ? flatten : to_a
+    new_self.length.times do
+      yield(new_self[i]) == true ? arr = arr.push(new_self[i]) : arr
+      i += 1
+    end
+    arr
+  end
```

Comment:
# my_each_with_index
 
- [ ] `my_each_with_index` when a block is given when self is a range returns the range itself after calling the given block once for each element in self

     Failure/Error: expect(range.my_each_with_index(&block)).to eq(range.each_with_index(&block))

       expected: 5..50
            got: [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50]

##

- [ ] `my_each_with_index` when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self

     Failure/Error: expect(my_each_with_index_output).to eq(each_with_index_output)

       expected: "Num: [:a, 1], idx: 0\nNum: [:b, 2], idx: 1\nNum: [:c, 3], idx: 2\nNum: [:d, 4], idx: 3\nNum: [:e, 5], idx: 4\n"
            got: "Num: a, idx: 0\nNum: 1, idx: 1\nNum: b, idx: 2\nNum: 2, idx: 3\nNum: c, idx: 4\nNum: 3, idx: 5\nNum: d, idx: 6\nNum: 4, idx: 7\nNum: e, idx: 8\nNum: 5, idx: 9\n"